### ISSUE TITLE
Support Windows' High Contrast Black (forced-colors)

### ISSUE THREAD

**Issue by chunhtai** [flutter/flutter#75883]
Windows firefox has unique behavior and does not works like other app. We should make sure flutter app works correctly under windows firefox browser in high contrast mode.

Test steps:

- turn on high contrast mode in [windows](https://support.microsoft.com/en-us/windows/use-high-contrast-mode-in-windows-10-fedc744c-90ac-69df-aed5-c8a90125e696)
- load flutter web using latest firefox and make sure things work correctly.

Internal issue: b/264456128

**Comment by goderbauer** [comment]
Implementation Idea: If we get the color values for the 15 [system-colors](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#system_colors) that this mode is defining we could create a material Theme that re-themes the application using only those colors. As a prototype, I've tried this with the flutter counter app and it seems to work well:

![image](https://user-images.githubusercontent.com/1227763/126004013-e6635cce-1e1f-443c-9cea-bbca7b4205f7.png)

Code at: https://gist.github.com/goderbauer/781509e473480e581bba8c33866c0fc6

There will be some customizations we will have to implement for some of the components: Currently, focused and disabled states of buttons for example are mainly expressed through changes in elevation and background color. Both are not visible in High Contrast Black. Instead, these modes are represented by changes in foreground color and/or outline color.

If the app defines custom colors somewhere not derived from the theme, these will not be mapped to one of the forced-colors colors with this approach. It would be the responsibility of the developer to use a color from the forced-color palette instead of the custom color if that mode is enabled. This is the inverse approach that browsers take with native websites where all colors are mapped to one of the forced-colors by default and developers then have to manually override that for custom colors that they do want to show up unchanged.

Additional Reading: 
* https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors
* https://blogs.windows.com/msedgedev/2020/09/17/styling-for-windows-high-contrast-with-new-standards-for-forced-colors/

**Comment by goderbauer** [comment]
Question is whether the web engine can pass the color values of the defined system-colors to the framework, e.g. in the [AccessibilityFeatures](https://master-api.flutter.dev/flutter/dart-ui/AccessibilityFeatures-class.html) object.

/cc @yjbanov

**Comment by goderbauer** [comment]
A user reports that `MediaQuery.highContrastOf(context)` always returns false on web. That probably also needs to change as part of this.

**Comment by yjbanov** [comment]
@nbayati Fixes this related issue - https://github.com/flutter/flutter/issues/104272 - so high contrast theme should be enabled. The part that's not implemented yet is mapping system colors of the Windows theme to the framework theme.

**Comment by jacobsimionato** [comment]
Changed this to P2 to match the priority of internal b/264456128 (P1).

**Comment by cbracken** [comment]
Removing the `[Web]` from the issue name since the samme issue affects the Windows desktop embedder as well.

**Comment by cbracken** [comment]
@nbayati are you specifically focusing on a fix for Flutter Web, or also planning to fix the desktop embedder? If only web, I can find someone to take a look at the desktop embedder side.

**Comment by nbayati** [comment]
@cbracken I've fixed the issue on web (https://github.com/flutter/flutter/issues/104272) to enable high contrast.
Forced-colors is a different feature that at some point we'll add in the web, but I'm not currently working on it. So I'm going to unassign myself and create a separate issue to track the development of supporting `forced-colors` feature.
Desktop is not really my area, so it'd be great if you can find someone to work on the desktop embedder.

**Comment by cbracken** [comment]
Thanks. High contrast is also already fixed on the desktop embedder, but forced colours is not. I had assumed this issue was only about the forced colours aspect.

**Comment by cbracken** [comment]
Self-assigning the portion of the forced colours issue required for the windows desktop embedder. 

#118853 was opened to cover web only it looks like.

**Comment by yaakovschectman** [comment]
I'm somewhat unclear on exactly what the goal of this issue is, specifically how the forced colors and developer-defined high contrast theme should contribute to the actual used colors. Do we want the system high contrast colors to take precedence over the provided high contrast theme, if there is one, or to be used as a default when none is provided?

**Comment by cbracken** [comment]
I've added a link in the issue description to the internal bug with more details.

**Comment by yaakovschectman** [comment]
https://github.com/flutter/engine/pull/39206 decides whether to use `highContrastTheme` or `highContrastDarkTheme` when in high contrast mode depending on the window color of the current theme, and can use some additional review at the moment.

**Comment by yjbanov** [comment]
Removed the web label from this issue. We have a separate one for web: https://github.com/flutter/flutter/issues/118853

**Comment by cbracken** [comment]
This is complete for Windows as of Yaakov's patch: https://github.com/flutter/engine/pull/39206
